### PR TITLE
Implement retry logic for `TestFile.copyFrom(URL)`

### PR DESCRIPTION
**Pull request by lacasseio**
This will help address flakiness on our infrastructure.

### CODE REVIEW

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java [hunk 1]
```diff
@@ -274,10 +274,16 @@ public void copyFrom(File target) {
...
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        int retries = 3;
+        while (retries-- > 0) {
+            try {
+                FileUtils.copyURLToFile(resource, this);
+                return;
+            } catch (IOException e) {
+                if (retries == 0) {
+                    throw new RuntimeException(e);
```
Comment (bmuschko) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: Let's throw a `UncheckedIOException` here. It's more specific.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java [hunk 2]
```diff
@@ -274,10 +274,16 @@ public void copyFrom(File target) {
     }
 
     public void copyFrom(URL resource) {
```
Comment (bmuschko) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: I'd love to see a test case for this method to ensure all situations are handled properly. Mocking or stubbing `FileUtils` might be a bit challenging though.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java (hunk 1 above)
Comment (lacasseio) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: Good idea. Done.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java (hunk 2 above)
Comment (lacasseio) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: I extracted the retry logic and use this two retry logic occurrences ([`WatchServiceRegistrar.watchDir`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/internal/filewatch/jdk7/WatchServiceRegistrar.java#L155-L191) and [`SocketConnection.writeWithNonBlockingRetry`](https://github.com/gradle/gradle/blob/master/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/SocketConnection.java#L256-L267)) as extended use case. Would you want me to adapt the extended use case to use `RetryUtil`?

File: subprojects/core/src/main/java/org/gradle/util/RetryUtil.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.util;
+
+import com.google.common.base.Function;
+import com.google.common.base.Throwables;
+import groovy.lang.Closure;
+
+public final class RetryUtil {
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/RetryUtil.java]: I'd move this class into the `internal-testing` module where we store testing infrastructure for unit tests.

File: subprojects/core/src/test/groovy/org/gradle/util/RetryUtilTest.groovy [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+package org.gradle.util
+
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class RetryUtilTest extends Specification {
+    @Unroll
+    def "exceed retry count of #retryCount throw last exception"() {
```
Comment (bmuschko) [inline subprojects/core/src/test/groovy/org/gradle/util/RetryUtilTest.groovy]: Missing "s" in "throw". Should be "throws".

File: subprojects/core/src/main/java/org/gradle/util/RetryUtil.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                lastException = e;
+            }
+        }
+
+        // Retry count exceeded, throwing last exception
+        Throwables.propagate(lastException);
+        return retryCount;
+    }
+
+    public static int retry(int retries, final Closure closure) {
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/RetryUtil.java]: I'd just make this a Groovy class. Then you can avoid declaring the first method. You can also use a `Closure` from Java as such `new Closure(null, null)` e.g. https://github.com/gradle/gradle/blob/master/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/AbstractTestDirectoryProvider.java#L93. Also, if you use Groovy you can provide default values for parameters.

File: subprojects/core/src/main/java/org/gradle/util/RetryUtil.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+package org.gradle.util;
+
+import com.google.common.base.Function;
+import com.google.common.base.Throwables;
+import groovy.lang.Closure;
+
+public final class RetryUtil {
+    private RetryUtil() {}
+
+    public static int retry(int retries, Function<Void, Void> function) {
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/RetryUtil.java]: The usage of this method looks cumbersome. For a more convenient solution see my comment below.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java [hunk 7]
```diff
@@ -16,6 +16,7 @@
 
 package org.gradle.test.fixtures.file;
 
+import com.google.common.base.Function;
```
Comment (bmuschko) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: This is probably unused now.

**Pull request merged**